envoy/upstream/upstream.h

```diff
@@ -311,7 +311,10 @@ class HostsPerLocality {
   /**
    * @return const std::vector<HostVector>& list of hosts organized per
    *         locality. The local locality is the first entry if
-   *         hasLocalLocality() is true.
+   *         hasLocalLocality() is true. All hosts within the same entry have the same locality
+   *         and all hosts with a given locality are in the same entry. With the exception of
+   *         the local locality entry (if present), all entries are sorted by locality with
+   *         those considered less by the LocalityLess comparator ordered earlier in the list.
```

Comment:
These assumptions are necessary for determining which localities overlap or don't overlap between two HostsPerLocality object. Without this, zone-aware routing has no way of safely comparing the local and upstream host set without additional assumptions about using the same zones in each host set.